File: 4.boggle_search.0.java
// java.util.* and java.util.streams.* have been imported for this problem.
// You don't need any other imports.
public static boolean validMove(char[][] board, boolean[][] free, int row, int col) {
    int minRow = 0;
    int minCol = 0;
    int maxRow = board.length;
    int maxCol = board[0].length;
    return minRow <= row 
        && row < maxRow 
        && minCol <= col 
        && col < maxCol 
        && free[row][col];
}

public static boolean matchChar(char[][] board, String word, int idx, int row, int col) {
    return word.charAt(idx) == board[row][col];
}

public static boolean helper(char[][] board, boolean[][] free, String word, int idx, int row, int col) {
    if (word.length() == idx) {
        return true;
    }
    if (validMove(board, free, row, col) && matchChar(board, word, idx, row, col)) {
        free[row][col] = false;
        idx+=1;
        boolean up = helper(board, free, word, idx, row-1, col);
        boolean left = helper(board, free, word, idx, row, col-1);
        boolean down = helper(board, free, word, idx, row+1, col);
        boolean right = helper(board, free, word, idx, row, col+1);
        idx-=1;
        free[row][col] = true;
        return up || left || down || right;
        
    }
    return false;
}

public static boolean boggleSearch(char[][] board, String word){
    if (word == null || board == null) return false;
    if (word.length() < 1 || board.length < 1 || board[0].length < 1) return false;
    boolean[][] free = new boolean[board.length][board[0].length];
    for (boolean[] row: free)
        Arrays.fill(row, true);
    for (int r=0; r<board.length; r++) {
        for (int c=0; c<board[0].length; c++) {
            if (helper(board, free, word, 0, r, c)) return true;
        }
    }
    return false;
}

File: 4.boggle_search.0.py
def can_move(board, unseen, cur_pos):
    r,c = cur_pos
    max_r,max_c = len(board), len(board[0])
    if 0 <= r < max_r and 0 <= c < max_c:
        return unseen[r][c]
    else:
        return False


def helper(board, unseen, cur_pos, remainder, results):
    if not remainder:
        result.append(True)
    r,c = cur_pos
    if remainder and board[r][c] == remainder[0]:
        unseen[r][c] = False
        new_remainder = remainder[1:]
        up = (r-1,c)
        if can_move(board, unseen, up):
            helper(board, unseen, up, new_remainder, results)
        down = (r+1,c)
        if can_move(board, unseen, down):
            helper(board, unseen, down, new_remainder, results)
        left = (r,c-1)
        if can_move(board, unseen, left):
            helper(board, unseen, left, new_remainder, results)
        right = (r,c+1)
        if can_move(board, unseen, right):
            helper(board, unseen, right, new_remainder, results)


def solution(board, word):
    results = list()
    max_r,max_c = len(board), len(board[0])
    for row in range(max_r):
        for col in range(max_r):
            if board[row][col] == word[0]:
                unseen = [[True for c in range(max_r)] for r in range(max_c)]
                start = (row,col)
                helper(board, unseen, start, word, results)
    return any(results)

